### ISSUE TITLE
Publishing default ASP.Net Core Web app template to azure doesn't modify the log location in web.config, resulting in no logs being produced #540

### ISSUE THREAD

**Issue by moozzyk** [aspnet/IISIntegration#167]
With the project created from templates the log path is set to be `.\logs\stdout`. If the user turns on logging but does not modify the path or delete the attribute the path will be used as is (publish-iis does not change path as it might have been set by the user). When publishing this project for Azure the path won't be correct and no logs are written. 

More details:
See: https://github.com/aspnet/Tooling-Internal/issues/540

Need to decide what we want to do.

**Comment by muratg** [comment]
@moozzyk  is the change in publish-iis command or the templates?

@Eilon @DamianEdwards do we want to fix this for RC2?

cc @danroth27 @glennc

**Comment by moozzyk** [comment]
I don't think it can be done in templates (or only in templates). When the application is being created you don't know whether it's going to be published for Azure or IIS. When publish-iis is invoked we know if we are publishing for Azure or not. This is when we have a chance to react. The question is how we should know whether to react. One option is to not put stdoutPath to the web.config by default (this would requires changes in templates) at all. In this case publish-iis will set the path accordingly. Another way is just to leave the templates alone as today and have publish-iis check the value - if the stdoutPath has the default value it would be overwritten otherwise it would be not touched to not overwrite something the user set. In this case the changes are needed only in publish-iis tool.

**Comment by Tratcher** [comment]
The other alternative is to have the templates set the log path to %LOG_PATH% like we do for the app and args. Then it publish-iis could check for this default and replace it for normal or azure publish.

**Comment by DamianEdwards** [comment]
If we can make a very simple fix (aka extremely-low-risk) to the templates ASAP then I'm OK with this for RC2. Otherwise, we should revisit this in RTM.

**Comment by moozzyk** [comment]
davidebbo commented 7 minutes ago • edited
Today, when git deploying an app to Azure, you end up with this in web.config:

<aspNetCore processPath="dotnet" arguments=".\AspNetCoreRC2.dll" stdoutLogEnabled="false" stdoutLogFile=".\logs\stdout" forwardWindowsAuthToken="false" />
To enable logging (including streaming), you need to change it to:

<aspNetCore processPath="dotnet" arguments=".\AspNetCoreRC2.dll" stdoutLogEnabled="true" stdoutLogFile="%home%\LogFiles\dotnetcore" forwardWindowsAuthToken="false" />
But Azure Web App already has an existing mechanism to enable logging, and it would be good to just honor it. It works as follows:

there is a file called settings.json in %home%\site\diagnostics
it contains something like this: {"AzureDriveEnabled":true,"AzureDriveTraceLevel":"Error"}
if AzureDriveEnabled is true, it logs to files in %home%\LogFiles\Applicatio

**Comment by davidebbo** [comment]
_Repasting my comment as your formatting got corrupted (should delete yours to avoid duplication)_

Today, when git deploying an app to Azure, you end up with this in `web.config`:

    <aspNetCore processPath="dotnet" arguments=".\AspNetCoreRC2.dll" stdoutLogEnabled="false" stdoutLogFile=".\logs\stdout" forwardWindowsAuthToken="false" />

To enable logging (including streaming), you need to change it to:

    <aspNetCore processPath="dotnet" arguments=".\AspNetCoreRC2.dll" stdoutLogEnabled="true" stdoutLogFile="%home%\LogFiles\dotnetcore" forwardWindowsAuthToken="false" />

But Azure Web App already has an existing mechanism to enable logging, and it would be good to just honor it. It works as follows:

- there is a file called `settings.json` in `%home%\site\diagnostics`
- it contains something like this: `{"AzureDriveEnabled":true,"AzureDriveTraceLevel":"Error"}`
- if `AzureDriveEnabled` is true, it logs to files in `%home%\LogFiles\Application`

**Comment by moozzyk** [comment]
The setting.json based logging mechanism uses .NET tracing - so it's tracing from the application. Here we are talking about IIS capturing whatever was written to standard output and dumping it to a file. Very useful for troubleshooting issues (e.g. startup) but probably not as much for application logging. Application can configure logging and decide to log elsewhere in which case these entries won't show up in stdOutLog. We will hover need to verify if the setting.json based Azure logging mechanism can be used to write application logs.

**Comment by moozzyk** [comment]
We decided to overwrite the outpath always to "\\?\%home%\LogFiles"  when publishing for Antares since the stdoutLog should be used for diagnostic purposes and the folder always exists.

**Comment by davidebbo** [comment]
`%home%\LogFiles\Application` might be a better choice. By default, when looking at log streams in VS, that's the only folder that gets looked at. So if you don't write there, most people will get confused (VS allows changing this, but it's an advanced setting).

**Comment by davidebbo** [comment]
@moozzyk please discuss with @glennc, as he's working on a separate logging provider which would likely conflict with this. His approach will grow better into supporting storage logging and more advanced features.

**Comment by moozzyk** [comment]
@davidebbo - as I said before - stdoutLogOutput is not for logging it is just for diagnosing (mostly startup) issues. It's especially useful to troubleshoot issues which happen even before we have a logger at all. Configuring a console logger and using stdoutLogOutput to log to a file is not a substitute for a file logger. 
As for the path - we have always used `%home%\LogFiles\` - feels a bit late to change this and I don't think it works out of the box. I just checked on one of my website on Azure and there is no `Application` subfolder in `LogFiles`:
```
> ls
D:\home\LogFiles
eventlog.xml
kudu
```
IIS does not create a folder for logging so you set stdoutLogEnabled and nothing is created and you are hitting this: https://github.com/projectkudu/kudu/issues/2014#issuecomment-220402231

**Comment by davidebbo** [comment]
I'm not sure I quite buy this argument. The fact is that by default apps are configured with:

    loggerFactory.AddConsole(Configuration.GetSection("Logging"));

Such that anything they log using `ILogger` will go there, for the lifetime of the console app (and not just the startup sequence).

Point is that if this is enabled and so is @glennc's logger, we will end up writing the same content under `D:\home\LogFiles` twice. And if someone turns on log streaming at the root (which is `D:\home\LogFiles`), every log line will be streamed twice.

**Comment by moozzyk** [comment]
Why would applications in production log to the Console? How do you even read it when you run in Antares (let's say AspNetCoreModule does not have stdoutLog feature)?

stdoutLog is a IIS/AspNetCoreModule feature and not AspNet's and you only turn it on when you need to diagnose things - it's unstructured and you may and will get additional spew not related to your application at all (e.g. when the application starts we Console.WriteLine some things, when there is a misplaced Console.WriteLine in the app it will get it). 

If you use stdoutLog only for diagnostic purposes you will get (especially when application starts) some additional content that logger can't get and you should not run your application with stdoutLog enabled otherwise (this is one of the reasons why it is disabled by default).

**Comment by davidebbo** [comment]
As long as users keep that off in Azure Web Apps and primarily use the new logger, I guess we're fine. And this will probably be the guidance, with the exception of diagnosing startup condition. If some *does* turn it on, then they will get duplicate stream, but hopefully this will be rare.

**Comment by moozzyk** [comment]
Agreed.

**Comment by moozzyk** [comment]
Fixed in 31a6a12

**Comment by GuardRex** [comment]
TL;DR: ANCM console logging :sparkles: just works :sparkles: in spite of its accidental existence and lack of support. However, it shouldn't be used probably because 3rd party logging implementations only require a little more effort to install and configure for a much longer list of logging features.

I used the "anti-pattern" for six months here successfully relying upon the ANCM-generated log with production and test apps. I would like to answer this for posterity and then make a completely different argument why it shouldn't be done:

> Why would applications in production log to the Console?

  1. .NET Logging is an abstraction for logging and not an implementation. Devs looking for really simple file-based logging are basically forced at :boy: :gun: to use a 3rd party implementation (e.g., Serilog, NLog, etc.) or to roll their own implementation, which is a huge PITA.
  2. Setup is crazy simple: (1) Add a dep on `Microsoft.Extensions.Logging.Console`, (2) Configure in `WebHostBuilder > .ConfigureLogging > options.AddConsole(LogLevel.Warning);` with `LogLevel.<level>` set based on the environment to whatever level you need to log (`Warning` for production, `Debug` for development), (3) `Console.WriteLine("...")` anywhere in the app to shoot logging messages to the console, thus to the file, and (4) Turn on the module logging `stdoutLogEnabled="true"` and `stdoutLogFile=".\logs\logfilename"`. (5) Done!
  3. It :sparkles: just works :sparkles:. I used this approach for six months and never had a problem. It's simple, secure, reliable, and efficient. It's all many need from a practical standpoint. If you're a small shop and just need one little log for each app to write a few lines into every so often or trap some data on exceptions, this works very well. No bells-and-whistles. Nothing fancy. It :sparkles: just works :sparkles:.
  4. I don't think the concerns hold-up from a practical point-of-view. (1) It doesn't work outside of ANCM: True! but `Console.WriteLine("...")` still logs to the Debug output on-screen, (2) It's not ASP.NET: So what? idk ... that's not a real reason not to do it IMO, (3) It's unstructured: So what? idk ... that's kind'a the point in that it's simple messages, (4) You'll get output in the log from `Console.WriteLine("...")`: That's the whole point ... that is how you log. (5) You'll get extra ANCM messages in your log: So what? That's a minimal inconvienance. Then, if there is an ANCM problem, it will get logged right along with the logging that you're doing in the same file.

This is a happy marriage between the simplicity of setting up console logging with the Logging infrastructure and the completely unrelated ANCM shelling out console to a file.

At this point, I would have suggested there should be another repo in .NET, maybe a part of Logging, maybe not, that would provide a basic logging implementation that works just like what the ANCM module is doing: Shelling out console to a file. For those that just want to `Console.WriteLine("...")` a few bits to a file here and there (or even go as far as I did and log exceptions across the entire app this way), it would work just like what I describe without having to rely upon the ANCM log, without having to use a 3rd party package, and without having to roll your own.

The real reason not to do this as far as I can see are not any of the reasons provided thus far, because they don't say "it isn't simple," "it isn't safe," and/or "it isn't practical." I think the best reason is the pay-for-play argument ... Logging implementations like Serilog, NLog, elmah.io, and Loggr are not much more difficult to setup and use. Because you get so much more bang-for-the-buck with really nice bells-and-whistles, devs should jump right into one of the 3rd party implementations. I still think MS could have done something for this, and I hope MS will, but until then, I see great logging happening with the 3rd party approaches (once you get past some dodgy documentation IMO).

**Comment by moozzyk** [comment]
https://github.com/aspnet/Logging/issues/441

**Comment by Sharpeli** [comment]
Is there any update on enabling logging to azure blob?

**Comment by GuardRex** [comment]
@Sharpeli There is no built-in logging implementation. There are implementations though with Azure sinks. Serilog has DocumentDB and Table Storage sinks. Log4Net has an Azure Blob piece.

https://github.com/serilog/serilog/wiki/Provided-Sinks
http://stemarie.github.io/log4net.Azure/

**Comment by Sharpeli** [comment]
@GuardRex  Thanks GuardRex, looks like the log4net is incompatible with .net core. Is there any workround to make the built-in “ILoggerFactory” or "Diagnostics.Trace" can log to azure blob for asp.net core app ?

**Comment by GuardRex** [comment]
@purekrome says on Slack to check out NLog. He says they use it quite a lot with great success, and it has an Azure Blob Storage sink.

- https://github.com/nlog/nlog/wiki/Targets
- http://stackoverflow.com/questions/23518078/store-nlog-generated-logs-to-azure-blob-storage-in-seperate-columns
- http://stackoverflow.com/questions/35058029/how-to-integrate-nlog-to-write-log-to-azure-streaming-log
https://aspnetcore.slack.com/messages/general/details/

**Comment by PureKrome** [comment]
Disclaimer: 
- I've never logged to Azure Blob storage before.
- I love using nLog
- I now use Serilog with an nLog sink (with various targets as required) :)

**Comment by Eilon** [comment]
I believe that's what https://github.com/aspnet/AzureIntegration/ does - it wires up logging along with Serilog and pushes log data to tables/blobs/wherever.

**Comment by GuardRex** [comment]
There isn't much in the way of guidance tho for the new package. cc/ @Rick-Anderson @danroth27 @spboyer

**Comment by Sharpeli** [comment]
Thanks all, I've found an simple trick for azure app ,see https://github.com/aspnet/Logging/tree/dev/src , just add the package "Microsoft.Extensions.Logging.AzureAppServices": "1.0.0-preview1-final" and update related dependencies,
 add the azure diagnostics in startup.cs like this : 
 loggerFactory.AddAzureWebAppDiagnostics();   // for default setting.
or 
loggerFactory.AddAzureWebAppDiagnostics(new AzureAppServicesDiagnosticsSettings( ...));   // add custom setting.
And enable the diagnostics log on azure, both logging on blob and file are work well. No need for any 
extra configuration. :)
